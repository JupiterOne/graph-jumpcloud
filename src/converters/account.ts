import { IntegrationRelationship } from "@jupiterone/jupiter-managed-integration-sdk";

import { JumpCloudOrg } from "../provider/types";
import { JumpCloudAccountEntity, JumpCloudUserEntity } from "../types";
import getTime from "../util/getTime";
import { consoleBaseUrl, createHasRelationship } from "./";

export const ACCOUNT_ENTITY_TYPE = "jumpcloud_account";
export const ACCOUNT_ENTITY_CLASS = ["Account", "Organization"];

export const ACCOUNT_USER_RELATIONSHIP_TYPE = "jumpcloud_account_has_user";

export function createAccountEntity(
  data: JumpCloudOrg,
): JumpCloudAccountEntity {
  return {
    _key: data.id || (data._id as string),
    _type: ACCOUNT_ENTITY_TYPE,
    _class: ACCOUNT_ENTITY_CLASS,
    id: data.id || (data._id as string),
    displayName: data.displayName as string,
    name: data.displayName as string,
    logoUrl: data.logoUrl as string,
    createdOn: getTime(data.created),
    webLink: consoleBaseUrl,
  };
}

export function createAccountUserRelationship(
  account: JumpCloudAccountEntity,
  user: JumpCloudUserEntity,
): IntegrationRelationship {
  return createHasRelationship(
    account._key,
    user._key,
    ACCOUNT_USER_RELATIONSHIP_TYPE,
  );
}
