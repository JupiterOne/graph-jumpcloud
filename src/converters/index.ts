import {
  EntityFromIntegration,
  RelationshipFromIntegration,
} from "@jupiterone/jupiter-managed-integration-sdk";

export * from "./constants";
export * from "./account";
export * from "./user";

export function createHasRelationships(
  fromEntity: EntityFromIntegration,
  toEntities: EntityFromIntegration[],
  relationshipType: string,
  relationshipProperties?: any,
) {
  const relationships: RelationshipFromIntegration[] = [];
  for (const e of toEntities) {
    relationships.push(
      createHasRelationship(
        fromEntity,
        e,
        relationshipType,
        relationshipProperties,
      ),
    );
  }
  return relationships;
}

export function createHasRelationship(
  fromEntity: EntityFromIntegration,
  toEntity: EntityFromIntegration,
  relationshipType: string,
  relationshipProperties?: any,
) {
  const relationship: RelationshipFromIntegration = {
    _key: `${fromEntity._key}|has|${toEntity._key}`,
    _type: relationshipType,
    _class: "HAS",
    _fromEntityKey: fromEntity._key,
    _toEntityKey: toEntity._key,
    displayName: "HAS",
    ...relationshipProperties,
  };

  return relationship;
}
