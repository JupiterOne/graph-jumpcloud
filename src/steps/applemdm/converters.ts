import { JumpCloudAppleDevice, JumpCloudAppleMDM } from '../../jumpcloud/types';
import { AppleEntities } from './constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
export function createAppleMDMEntity(data: JumpCloudAppleMDM) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id,
        _type: AppleEntities.APPLE_MDM._type,
        _class: AppleEntities.APPLE_MDM._class,
        id: data.id,
        displayName: data.name,
        depServerTokenState: data.depServerTokenState,
        apnsCertExpiry: data.apnsCertExpiry,
        appleCertSerialNumber: data.appleCertSerialNumber,
      },
    },
  });
}
export function createAppleMDMDevice(data: JumpCloudAppleDevice) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id,
        _type: AppleEntities.APPLE_DEVICE._type,
        _class: AppleEntities.APPLE_DEVICE._class,
        id: data.id,
        serial: data.serialNumber,
        deviceId: data.udid,
        osVersion: data.osVersion,
        depRegistered: data.depRegistered,
        enrolled: data.enrolled,
        createdAt: parseTimePropertyValue(data.createdAt),
        hasActivationLockBypassCodes: data.hasActivationLockBypassCodes,
        isSupervised: data.deviceInformation.isSupervised,
        deviceCapacity: data.deviceInformation.deviceCapacity,
        name: data.deviceInformation.deviceName,
        macAddress: data.deviceInformation.wifiMac,
        category: data.deviceInformation.modelName,
        model: data.deviceInformation.modelName,
        make: 'Apple',
      },
    },
  });
}
