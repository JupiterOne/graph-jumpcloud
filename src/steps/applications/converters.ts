import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { JumpCloudApplication } from '../../jumpcloud/types';
import { getConsoleUrl } from '../../jumpcloud/url';
import { ApplicationEntities } from './constants';

export function createApplicationEntity(data: JumpCloudApplication) {
  const appId = data.id || (data._id as string);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: appId,
        _type: ApplicationEntities.APPLICATION._type,
        _class: ApplicationEntities.APPLICATION._class,
        id: data.id as string,
        name: data.name as string,
        displayName: data.displayName as string,
        description: data.description,
        createdOn: parseTimePropertyValue(data.created),
        ssoUrl: data.ssoUrl,
        webLink: getConsoleUrl(`/#/sso/${data.id}/details`),
      },
    },
  });
}
