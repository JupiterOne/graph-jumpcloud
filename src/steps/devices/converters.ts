import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { JumpCloudDevice } from '../../jumpcloud/types';
import { DeviceEntities } from './constants';

export function createDeviceEntity(data: JumpCloudDevice) {
  const deviceId = data.id || (data._id as string);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: deviceId,
        _type: DeviceEntities.DEVICE._type,
        _class: DeviceEntities.DEVICE._class,
        id: data.id,
        name: data.deviceInformation.deviceName,
        displayName: data.deviceInformation.deviceName,
        createdOn: parseTimePropertyValue(data.created),
      },
    },
  });
}
