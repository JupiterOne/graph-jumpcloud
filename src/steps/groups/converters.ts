import { createIntegrationEntity, parseTimePropertyValue } from "@jupiterone/integration-sdk-core";
import { JumpCloudGroup } from "../../jumpcloud/types";
import { getConsoleUrl } from "../../jumpcloud/url";
import { GroupEntities } from "./constants";

export function createGroupEntity(
  data: JumpCloudGroup
) {
  const userId = data.id || (data._id as string);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: userId,
        _type: GroupEntities.GROUP._type,
        _class: GroupEntities.GROUP._class,
        id: data.id as string,
        name: data.name as string,
        displayName: data.name as string,
        createdOn: parseTimePropertyValue(data.created),
        attributes: (data.attributes && JSON.stringify(data.attributes)) || undefined,
        webLink: getConsoleUrl(`/#/groups/user/${data.id}/details`)
      }
    }
  });
}
