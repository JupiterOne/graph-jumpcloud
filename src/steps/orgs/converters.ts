import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { JumpCloudOrg } from '../../jumpcloud/types';
import { consoleBaseUrl } from '../../jumpcloud/url';
import { OrgEntities } from './constants';

export function createOrgEntity(data: JumpCloudOrg) {
  const orgId = data.id || (data._id as string);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: orgId,
        _type: OrgEntities.ORG._type,
        _class: OrgEntities.ORG._class,
        id: orgId,
        displayName: data.displayName || undefined,
        name: data.displayName || orgId,
        logoUrl: data.logoUrl || undefined,
        createdOn: parseTimePropertyValue(data.created),
        webLink: consoleBaseUrl,
      },
    },
  });
}
