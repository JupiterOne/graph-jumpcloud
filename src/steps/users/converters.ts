import { createIntegrationEntity, parseTimePropertyValue } from "@jupiterone/integration-sdk-core";
import { JumpCloudUser } from "../../jumpcloud/types";
import { getConsoleUrl } from "../../jumpcloud/url";
import { UserEntities } from "./constants";

export function createUserEntity(
  data: JumpCloudUser
) {
  const userId = data.id || (data._id as string);

  return createIntegrationEntity({
    entityData: {
      source: {
        ...data,
        tags: [],
        userTags: data.tags,
      },
      assign: {
        _key: userId,
        _type: UserEntities.USER._type,
        _class: UserEntities.USER._class,
        id: userId,
        displayName: data.username as string,
        name: `${data.firstname} ${data.lastname}`,
        username: data.username,
        active: !!data.activated,
        suspended: !!data.suspended,
        createdOn: parseTimePropertyValue(data.created)!,
        firstName: data.firstname,
        lastName: data.lastname,
        login: data.username,
        email: data.email,
        employeeType: data.employeeType,
        employeeId: data.employeeIdentifier || undefined,
        company: data.company,
        department: data.department,
        jobTitle: data.jobTitle,
        mfaEnabled: !!data.totp_enabled,
        mfaExclusion: !!data.mfa!.exclusion,
        mfaExclusionUntil: parseTimePropertyValue(data.mfa!.exclusionUntil),
        mfaConfigured: !!data.mfa!.configured,
        externallyManaged: !!data.externally_managed,
        ldapBindingUser: !!data.ldap_binding_user,
        sambaServiceUser: !!data.samba_service_user,
        sudo: !!data.sudo,
        webLink: getConsoleUrl(`/#/users/${userId}/details`)
      }
    }
  });
}
