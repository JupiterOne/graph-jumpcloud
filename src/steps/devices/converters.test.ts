import { createMockDevice } from '../../../test/mocks';
import { createDeviceEntity } from './converters';

describe('#createDeviceEntity', () => {
  test('should convert data', () => {
    expect(createDeviceEntity(createMockDevice())).toMatchSnapshot();
  });
});
