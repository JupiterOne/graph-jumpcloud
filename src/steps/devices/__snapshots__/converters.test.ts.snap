// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createDeviceEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Device",
  ],
  "_key": "mockid123",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "createdAt": "2021-10-29T17:59:00.448Z",
        "depRegistered": true,
        "deviceInformation": Object {
          "activationLockAllowedWhileSupervised": true,
          "availableDeviceCapacity": 3,
          "deviceCapacity": 2,
          "deviceName": "mockDevice",
          "iccid": "mockiccid",
          "imei": "mockimei",
          "isActivationLockEnabled": true,
          "isSupervised": true,
          "modelName": "mockmodelname",
          "subscriberCarrierNetwork": "mockcarriernetworkname",
          "wifiMac": "aa:bb:cc:dd:ee:ff",
        },
        "enrolled": true,
        "hasActivationLockBypassCodes": true,
        "id": "mockid123",
        "osVersion": "mockmackosversion12",
        "securityInfo": Object {
          "enrolledViaDep": true,
          "isActivationLockManageable": true,
          "isUserEnrollment": true,
          "passcodePresent": true,
          "userApprovedEnrollment": true,
        },
        "serialNumber": "123456abc",
        "udid": "mockudid",
      },
    },
  ],
  "_type": "jumpcloud_device",
  "createdOn": 1635530340448,
  "displayName": "mockDevice",
  "id": "mockid123",
  "name": "mockDevice",
  "osVersion": "mockmackosversion12",
}
`;
